sips/sip-xyz/sip-xyz-bitwise-ops.md

```diff
@@ -0,0 +1,206 @@
+
+- Bitwise Xor (`^`)
+- Bitwise And (`&`)
+- Bitwise Or (`|`)
+- Bitwise Not (`~`)
+- Binary Left Shift (`<<`)
+- Binary Right Shift (`>>`)
+
+# License and Copyright
+
+This SIP is made available under the terms of the Creative Commons CC0 1.0
+Universal license, available at
+https://creativecommons.org/publicdomain/zero/1.0/ This SIP's copyright is held
+by the Stacks Open Internet Foundation.
+
+# Introduction
+
+Bitwise operations are common in other programming languages. Common algorithms,
+including many used in encryption, or the ability to set and check flags in a
+bit field for example, would be much more difficult to implement without the use
+of these operations. When executing a contract using these operations, the
+common hardware on which miners and nodes are likely to be running can all
+perform these operations very efficiently -- these are typically single cycle
+operations. Note that the addition of these bitwise operations commits the
+Clarity VM to using 2's complement to represent integers.
+
+# Specification
+
+## Bitwise Xor (`^`)
+
+`(^ i1 i2...)`
+
+- **Inputs:** int, ... | uint, ...
+- **Output:** int | uint
+
+Returns the result of bitwise exclusive or'ing a variable number of integer
+inputs.
+
+### Examples
+
+```
+(^ 1 2) ;; Returns 3
+(^ 120 280) ;; Returns 352
+(^ -128 64) ;; Returns -64
+(^ u24 u4) ;; Returns u28
+(^ 1 2 4 -1) ;; Returns -8
+```
+
+## Bitwise And (`&`)
+
+`(& i1 i2...)`
+
+- **Inputs:** int, ... | uint, ...
+- **Output:** int | uint
+
+Returns the result of bitwise and'ing a variable number of integer inputs.
+
+### Examples
+
+```
+(& 24 16) ;; Returns 16
+(& 28 24 -1) ;; Returns 24
+(& u24 u16) ;; Returns u16
+(& -128 -64) ;; Returns -128
+(& 28 24 -1) ;; Returns 24
+```
+
+## Bitwise Or (`|`)
+
+`(& i1 i2...)`
+
+- **Inputs:** int, ... | uint, ...
+- **Outputs:** int | uint
+
+Returns the result of bitwise inclusive or'ing a variable number of integer
+inputs.
+
+### Examples
+
+```
+(| 4 8) ;; Returns 12
+(| 1 2 4) ;; Returns 7
+(| 64 -32 -16) ;; Returns -16
+(| u2 u4 u32) ;; Returns u38
+```
+
+## Bitwise Not (`~`)
+
+`(~ i1)`
+
+- **Inputs:** int | uint
+- **Output:** int | uint
+
+Returns the one's compliment (sometimes also called the bitwise compliment or
+not operator) of `i1`, effectively reversing the bits in `i1`.
+
+In other words, every bit that is `1` in `ì1` will be `0` in the result.
+Conversely, every bit that is `0` in `i1` will be `1` in the result.
+
+### Examples
+
+```
+(~ 3) ;; Returns -4
+(~ u128) ;; Returns u340282366920938463463374607431768211327
+(~ 128) ;; Returns -129
+(~ -128) ;; Returns 127
+```
+
+## Bitwise Left Shift (`<<`)
+
+`(<< i1 shamt)`
+
+- **Inputs:** int, uint | uint, uint
+- **Outputs:** int | uint
+
+Shifts all bits in `i1` to the left by the number of places specified in `shamt`
+modulo 128 (the bit width of Clarity integers). New bits are filled with zeros.
+When `i1` is a `uint` (unsigned), this is a logical shift. When `i1` is an `int`
```

Comment:
On further testing, this text is wrong.  The sign bit is _not_ preserved for `int`, and that's also consistent with an arithmetic left shift.

Given that we aren't doing any arithmetic overflows here, it might be prudent to mention up front that this is a deliberate choice.  If the developer wants arithmetic errors, they can use `*`, `/`, and `pow`.

Reply:
> On further testing, this text is wrong. The sign bit is not preserved for int, and that's also consistent with an arithmetic left shift.
Yes, I noticed this too. Will update.

Reply:
Updated in [a2cf4f6](https://github.com/stacksgov/sips/pull/106/commits/a2cf4f6bdb86da855039a7fc87f4c56b08c0a43f)